test/unit/core/type_traits/function_test.cpp

```diff
@@ -35,3 +35,245 @@ TEST(function_traits, argument_type_at)
+
+    using fn_void_0_param_t = decltype(fn_void_0_param);
+    using fn_bool_2_param_t = decltype(fn_bool_2_param);
+
+    using fn_void_0_param_const_t = decltype(fn_void_0_param_const);
+    using fn_bool_2_param_const_t = decltype(fn_bool_2_param_const);
+
+    using fn_void_0_param_noexcept_t = decltype(fn_void_0_param_noexcept);
+    using fn_bool_2_param_noexcept_t = decltype(fn_bool_2_param_noexcept);
+
+    using fn_void_0_param_lvalue_ref_t = decltype(fn_void_0_param) &;
+    using fn_bool_2_param_lvalue_ref_t = decltype(fn_bool_2_param) &;
+
+    using fn_void_0_param_rvalue_ref_t = decltype(fn_void_0_param) &&;
+    using fn_bool_2_param_rvalue_ref_t = decltype(fn_bool_2_param) &&;
+
+    using fn_void_0_param_complex_t = decltype(fn_void_0_param_complex) &;
+    using fn_bool_2_param_complex_t = decltype(fn_bool_2_param_complex) &&;
+
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_const_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_const_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_noexcept_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_noexcept_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_lvalue_ref_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_lvalue_ref_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_rvalue_ref_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_rvalue_ref_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_void_0_param_complex_t>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_bool_2_param_complex_t>::value);
+}
+
+struct fn_object_simple
+{
+    auto operator()(int)
+    {
+        return true;
+    }
+};
+
+struct fn_object_const_lvalue_ref
+{
+    auto operator()(int) &
+    {
+        return true;
+    }
+};
+
+struct fn_object_const_complex
+{
+    auto operator()(int) const & noexcept(std::is_nothrow_default_constructible_v<bool>)
+    {
+        return true;
+    }
+};
+
+TEST(is_function, function_object)
+{
+    EXPECT_TRUE(seqan3::is_function<fn_object_simple>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_lvalue_ref>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_complex>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_simple &>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_lvalue_ref &>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_complex &>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_simple &&>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_lvalue_ref &&>::value);
+    EXPECT_TRUE(seqan3::is_function<fn_object_const_complex &&>::value);
+}
+
+struct fn_object_rvalue_ref
+{
+    auto operator()(int) &&
+    {
+        return true;
+    }
+};
+
+TEST(is_function, false_functions)
+{
+    auto generic_fn = [] (auto x) { return x; };
```

Comment:
Can you add a function object with a template argument.

Reply:
a class template or a generic call operator?